From d5e4fecdced000b17027dc530591972ece3ee102 Mon Sep 17 00:00:00 2001
From: Yadu MG <quic_ymg@quicinc.com>
Date: Sun, 2 Mar 2025 12:20:01 +0530
Subject: [PATCH] PENDING: firmware: qcom-scm: Add Mutex locking mechanism for
 SCM driver

Currently SMCInvoke call paralleization is supported with wait queue
and similar support is not enabled for the PIL subsystem calls. For
PIL subysystem calls when occurs in parallel, leads to deadlock in
TZ/HYP. To serialize the calls, mutex locking is introduced to avoid
the behaviour.

Change-Id: Ib1bdbc172aca55c08804d158ed1c5c4009e44858
Signed-off-by: Yadu MG <quic_ymg@quicinc.com>
---
 drivers/firmware/qcom_scm-smc.c   | 8 ++++++++
 drivers/firmware/qcom_scm_addon.c | 3 +++
 2 files changed, 11 insertions(+)

diff --git a/drivers/firmware/qcom_scm-smc.c b/drivers/firmware/qcom_scm-smc.c
index 78bcb7eb491a..1674dde1e349 100644
--- a/drivers/firmware/qcom_scm-smc.c
+++ b/drivers/firmware/qcom_scm-smc.c
@@ -25,6 +25,8 @@ struct arm_smccc_args {
 	unsigned long args[8];
 };
 
+static DEFINE_MUTEX(qcom_scm_lock);
+
 #define QCOM_SCM_EBUSY_WAIT_MS 30
 #define QCOM_SCM_EBUSY_MAX_RETRY 20
 
@@ -154,12 +156,18 @@ static int __scm_smc_do(struct device *dev, struct arm_smccc_args *smc,
 	}
 
 	do {
+		if(!skip_mutex)
+			mutex_lock(&qcom_scm_lock);
+
 		down(&qcom_scm_sem_lock);
 
 		ret = __scm_smc_do_quirk_handle_waitq(dev, smc, res, skip_mutex);
 
 		up(&qcom_scm_sem_lock);
 
+		if(!skip_mutex)
+			mutex_unlock(&qcom_scm_lock);
+
 		if (ret)
 			return ret;
 
diff --git a/drivers/firmware/qcom_scm_addon.c b/drivers/firmware/qcom_scm_addon.c
index 23f2af254a99..4fe7816d26a3 100644
--- a/drivers/firmware/qcom_scm_addon.c
+++ b/drivers/firmware/qcom_scm_addon.c
@@ -380,6 +380,7 @@ int qcom_scm_invoke_smc(phys_addr_t in_buf, size_t in_buf_size,
 		.args[3] = out_buf_size,
 		.arginfo = QCOM_SCM_ARGS(4, QCOM_SCM_RW, QCOM_SCM_VAL,
 					QCOM_SCM_RW, QCOM_SCM_VAL),
+		.skip_mutex = true,
 	};
 	struct qcom_scm_res res;
 
@@ -413,6 +414,7 @@ int qcom_scm_invoke_smc_legacy(phys_addr_t in_buf, size_t in_buf_size,
 		.args[3] = out_buf_size,
 		.arginfo = QCOM_SCM_ARGS(4, QCOM_SCM_RW, QCOM_SCM_VAL,
 					QCOM_SCM_RW, QCOM_SCM_VAL),
+		.skip_mutex = true,
 	};
 
 	struct qcom_scm_res res;
@@ -444,6 +446,7 @@ int qcom_scm_invoke_callback_response(phys_addr_t out_buf,
 		.args[0] = out_buf,
 		.args[1] = out_buf_size,
 		.arginfo = QCOM_SCM_ARGS(2, QCOM_SCM_RW, QCOM_SCM_VAL),
+		.skip_mutex = true,
 	};
 	struct qcom_scm_res res;
 
-- 
2.25.1

